File: 2019/day10.py
import math
import sys

# 3,4 with 8

t1 = """\
.#..#
.....
#####
....#
...##
"""

# 5,6 with 33

t2 = """\
#..#.#....
..#######.
.#.#.###..
.#..#.....
..#....#.#
#..#....#.
.##.#..###
##...#..#.
.#....####
"""


# 1,2 with 35

t3 = """\
#.#...#.#.
.###....#.
.#....#...
##.#.#.#.#
....#.#.#.
.##..###.#
..#...##..
..##....##
......#...
.####.###.
"""

# 6,3 with 41

t4 = """\
.#..#..###
####.###.#
....###.#.
..###.##.#
##.##.#.#.
....###..#
..#.#..#.#
#..#.#.###
.##...##.#
.....#.#..
"""

# 11,13 with 210

t5 = """\
.#..##.###...#######
##.############..##.
.#.######.########.#
.###.#######.####.#.
#####.##.#.##.###.##
..#####..#.#########
####################
#.####....###.#.#.##
##.#################
#####.##.###..####..
..######..##.#######
####.##.####...##..#
.#####..#.######.###
##...#.##########...
#.##########.#######
.####.#.###.###.#.##
....##.##.###..#####
.#.#.###########.###
#.#.#.#####.####.###
###.##.####.##.#..##
"""

def read(t):
    if isinstance(t,str):
        t = t.splitlines()
    else:
        t = t.readlines()
    return [list(ln.strip()) for ln in t]

def gcd(x,y):
    while y:
        x,y = y,x%y
    return x

def visible( graph, x0, y0, x1, y1 ):
    # Get the slope of the line.
    dx = x1 - x0
    dy = y1 - y0
    
    # Compute the GCD.  If they are relatively prime, then this star is visible.
    div = gcd(abs(dx),abs(dy))
    if div == 1:
        return True

    # Compute the steps at which stars will interfere.
    sx = dx // div
    sy = dy // div

    # until we reach the target (which we always will), check for blocks.
    x0 += sx
    y0 += sy
    while (x0,y0) != (x1,y1):
        if graph[y0][x0] == '#':
            return False
        x0 += sx
        y0 += sy
    return True


def check( graph, tx, ty ):
#    print( "Checking", tx, ty )
    dim = len(graph)
    found = set()
    for y1 in range(dim):
        for x1 in range(dim):
            if (x1,y1) == (tx,ty):
                continue
            if graph[y1][x1] != '#':
                continue
            if visible( graph, tx,ty, x1,y1 ):
                found.add( (x1, y1) )
#    print( len(found) )
    return found



def scan(graph):
    maxcount = 0
    for y,ln in enumerate(graph):
        for x,cell in enumerate(ln):
#            print( x, y, ln, cell )
            if cell == '.':
                continue
            count = len(check( graph, x, y ))
            if count > maxcount:
                maxcount = count
                maxelem = (x,y)
    print( maxelem, "sees", maxcount )
    return maxcount, maxelem


def laser(graph,x0,y0):
    remains = 200
    while remains:
        found = check(graph,x0,y0)
# If we don't catch them all in one round, eliminate these.
        if len(found) < remains:
            for x1,y1 in found:
                graph[y1][x1] = '.'
            remains -= len(found)
            continue
        break

# Compute the angle from north.
    data = [
        ((math.pi/4 - math.atan2((x1-x0),(y1-y0))), x1, y1)
        for x1,y1 in found
    ]

# Sort by the angle, pick the 200th.
    data.sort()
    print( "Number 200 is", data[remains-1] )
    return data[remains-1][1] * 100 + data[remains-1][2]

    

def part1():
    scan(read(t1))
    scan(read(t2))
    scan(read(t3))
    scan(read(t4))
    scan(read(t5))
    print( "Part 1: ", scan(read(open('day10.txt')))[0])

# Answer was 26, 29.
        
def part2():
    best = scan(read(t5))[1]
    laser( read(t5), *best )
    best = scan(read(open('day10.txt')))[1]
    print( "\nPart 2: ", laser(read(open('day10.txt')), *best) )

part1()
print()
part2()

File: 2019/day10.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <cmath>
#include <vector>
#include <tuple>
#include <set>
#include <algorithm>

#ifndef M_PI
#define M_PI 3.14159265358979
#endif

// 3,4 with 8

const char * t1 = "\
.#..#\n\
.....\n\
#####\n\
....#\n\
...##\n\
";

// 5,6 with 33

const char * t2 = "\
#..#.#....\n\
..#######.\n\
.#.#.###..\n\
.#..#.....\n\
..#....#.#\n\
#..#....#.\n\
.##.#..###\n\
##...#..#.\n\
.#....####\n\
";


// 1,2 with 35

const char * t3 = "\
#.#...#.#.\n\
.###....#.\n\
.#....#...\n\
##.#.#.#.#\n\
....#.#.#.\n\
.##..###.#\n\
..#...##..\n\
..##....##\n\
......#...\n\
.####.###.\n\
";

// 6,3 with 41

const char * t4 = "\
.#..#..###\n\
####.###.#\n\
....###.#.\n\
..###.##.#\n\
##.##.#.#.\n\
....###..#\n\
..#.#..#.#\n\
#..#.#.###\n\
.##...##.#\n\
.....#.#..\n\
";

// 11,13 with 210

const char * t5 = "\
.#..##.###...#######\n\
##.############..##.\n\
.#.######.########.#\n\
.###.#######.####.#.\n\
#####.##.#.##.###.##\n\
..#####..#.#########\n\
####################\n\
#.####....###.#.#.##\n\
##.#################\n\
#####.##.###..####..\n\
..######..##.#######\n\
####.##.####...##..#\n\
.#####..#.######.###\n\
##...#.##########...\n\
#.##########.#######\n\
.####.#.###.###.#.##\n\
....##.##.###..#####\n\
.#.#.###########.###\n\
#.#.#.#####.####.###\n\
###.##.####.##.#..##\n\
";


struct Point {
    int x;
    int y;
    Point( )
        : x(0)
        , y(0)
        {}
    Point( int _x, int _y )
        : x(_x)
        , y(_y)
        {}

    bool operator< (const Point & other) const
    {
        return x < other.x || y < other.y;
    }

    bool operator== (const Point & other) const
    {
        return x==other.x && y==other.y;
    }

    bool operator!= (const Point & other) const
    {
        return x!=other.x || y!=other.y;
    }

    Point operator+( const Point & other )
    {
        return Point( x+other.x, y+other.y );
    }

    Point& operator+=( const Point & other )
    {
        x += other.x;
        y += other.y;
        return *this;
    }
};


//typedef std::vector<std::vector<char>> spacemap_t;

struct spacemap_t : public std::vector<std::vector<char>> 
{
    // This seems silly.

    std::vector<char> & operator[]( int y )
    {
        return at(y);
    }
    const std::vector<char> & operator[]( int y ) const
    {
        return at(y);
    }
    char & operator[]( const Point & pt )
    {
        return at(pt.y)[pt.x];
    }
    char operator[]( const Point & pt ) const
    {
        return at(pt.y)[pt.x];
    }
};


spacemap_t read( std::istream  && t )
{
    spacemap_t map;
    for( std::string ln; getline( t, ln); )
    {
        map.emplace_back( ln.size() );
        std::copy( 
            ln.begin(),
            ln.end(),
            map.back().begin()
        );
    }
    return map;
}


spacemap_t read( const char * t )
{
    return read( std::istringstream(t) );
}


int gcd( int x, int y )
{
    while( y )
    {
        int t = y;
        y = x % y;
        x = t;
    }
    return x;
}


bool visible( const spacemap_t & graph, Point p0, Point p1 )
{
    // Get the slope of the line.

    int dx = p1.x - p0.x;
    int dy = p1.y - p0.y;
    
    // Compute the GCD.  If they are relatively prime, then this star 
    // is visible.

    int div = gcd(abs(dx),abs(dy));
    if( div == 1 )
        return true;

    // Compute the steps at which stars will interfere.

    Point step( dx / div, dy / div );

    // Until we reach the target (which we always will), check for blocks.

    for( p0 += step; p0 != p1; p0 += step )
        if( graph[p0] == '#' )
            return false;

    return true;
}


std::set<Point> check( const spacemap_t & graph, Point(target) )
{
//    std::cout << "Checking " << target.x << " " <<  target.y << "\n";
    int dim = graph.size();
    std::set<Point> found;
    for( int y1=0; y1 < dim; y1++ )
    {
        for( int x1=0; x1 < dim; x1++ )
        {
            Point pt(x1,y1);
            if( pt == target )
                continue;
            if( graph[pt] != '#' )
                continue;
            if( visible( graph, target, pt ) )
            {
//                std::cout << "   " << pt.x << "," << pt.y << " visible\n";
                found.insert( pt );
            }
        }
    }
    return found;
}



std::pair<int,Point> scan(const spacemap_t & graph) 
{
    int maxcount = 0;
    Point maxelem(0,0);
    for( int y = 0; y < graph.size(); y++ )
    {
        auto ln = graph[y];
        for( int x = 0; x < ln.size(); x++ )
        {
            char cell = ln[x];
//            print( x, y, ln, cell )
            if( cell == '.' )
                continue;
            int count = check( graph, Point(x, y) ).size();
            if( count > maxcount )
            {
                maxcount = count;
                maxelem = Point(x,y);
            }
        }
    }
    std::cout << maxelem.x << "," << maxelem.y << " sees " << maxcount << "\n";
    return std::make_pair( maxcount, maxelem );
}


double angle( Point & a, Point & b )
{
    return M_PI/4 - atan2((a.x-b.x),(a.y-b.y));
}


int laser(spacemap_t graph, Point target)
{
    int remains = 200;
    std::set<Point> found;

    // If we don't catch them all in one round, eliminate this whole set.

    while( (found = check(graph,target)).size() < remains )
    {
        for( auto pt : found )
            graph[pt] = '.';
        remains -= found.size();
    }

    // Convert to a vector.

    std::vector<Point> data;
    data.resize( found.size() );
    std::copy( found.begin(), found.end(), data.begin() );

    std::sort(
        data.begin(),
        data.end(),
        [&target](Point & a, Point & b) {
            return angle(a,target) < angle(b,target);
        } 
    );

    std::cout << "Number 200 is " 
        << data[remains-1].x << ","
        << data[remains-1].y << "\n";

    return data[remains-1].x * 100 + data[remains-1].y;
}


void part1()
{
    scan(read(t1));
    scan(read(t2));
    scan(read(t3));
    scan(read(t4));
    scan(read(t5));
    std::cout 
        << "\nPart 1: " 
        << scan(read(std::ifstream("day10.txt"))).first << "\n";
}

// Answer was 26, 29.
        
void part2()
{
    Point best = scan(read(t5)).second;
    std::cout << laser( read(t5), best ) << "\n";

    // Find the point with the most visible asteroids.
    best = scan(read(std::ifstream("day10.txt"))).second;

    // Find the 200th blasted asteroid.
    std::cout 
        << "\nPart 2: " 
        << laser(read(std::ifstream("day10.txt")), best)  << "\n";
}

int main()
{
    part1();
    std::cout << std::endl;
    part2();
}
